mds: add ceph.dir.bal.mask vxattr for MDS Balancer (#52373)
ceph/ceph: doc/cephfs/multimds.rst

.. prompt:: bash #

    ceph fs set <fs_name> bal_rank_mask 0x0

Dynamically partitionning directory trees with vxattr on a specific ranks
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The ``bal_rank_mask`` option can be overridden by configuring the ``ceph.dir.bal.mask`` vxattr.

Author: Fixed

> Notably, the ``ceph.dir.bal.mask`` offers the capability to segregate particular subdirectories

Author: Fixed.

into dedicated groups of MDS ranks, whereas bal_rank_mask only allows isolation

> at the filesystem level, such as the root directory ``/``. It serves as a valuable

Author: Fixed.

tool for finely tuning and enhancing MDS performance in various scenarios.

One scenario is when dealing with a large directory that proves challenging
for a single MDS rank to handle efficiently. For example, the ``/home`` directory
contains many directories for multiple users. In most cases, static pinning is
simply used to distribute subdirectories of /home across multiple ranks. However,

> the ``/home/images`` directory presents a unique situation with subdirectories 0 to 99,

Author: I think ```/usr``` is better. However, I don't think it will mount cephfs on ```/usr```. What about examples like ```/data``` or ```/cephfs``` directories?

> the ``/home/images`` directory presents a unique situation with subdirectories 0 to 99,

Reviewer: Well, how do you plan to use it? I think `/usr` could make sense in the context of a backing store of a large software respository file system like cvmfs. It just requires exposition to explain.

> the ``/home/images`` directory presents a unique situation with subdirectories 0 to 99,

Author: We need to find a good use case. We had an issue with saving large files at https://laion.ai/blog/laion-400-open-dataset. In fact, cephfs users mount and store file systems in any directory. It is difficult to determine a general case because the user stores it in any directory they want. For example: ```/home/$user/laion-5b```, ```/laion-5b``` , or ```/mnt/laion-5b```. cvmfs is also a good example. I'll think about it.

> the ``/home/images`` directory presents a unique situation with subdirectories 0 to 99,

Author: There are no good examples. cvmfs also uses the /cvmfs directory. some helpful explanation is added to the note.

https://github.com/ceph/ceph/compare/9b04cd83fb615464888de665b9933d37724d4a28..ac1ab14f09a789361f9a5c72e448ae85f5b55fa8#diff-cc9cf0d49df0c5644ecb9c98d1f7ac1d307b05837bfdabc6761c65691c6f8dc1R344

each containing large number of files, often exceeding a million files. With
this example with ``/home/images``, processing them in one MDS is inappropriate
due to insufficient MDS resources like MDS cache memory. Therefore,
the ceph.dir.bal.mask can dynamically balance the workload of this large
directory within a specific subset of ranks.

The ``ceph.dir.bal.mask`` can also be a useful option for fine-tunning performance,
particularly in scenarios involving large directories like ``/home/images``
and ``/home/backups`` within the file system.

> While the mdsmap’s ``bal_rank_mask`` isolates the entire ``/`` directory to

Author: Fixed.

specific ranks, it can affect performance due to each other's migration overhead.
For example, if the mdsmap’s ``bal_rank_mask`` is set to ``0xf`` and large directories
like ``/home/images`` and ``/home/backups`` exist, the load on ``/home/images``
instantaneously increases and metadata distribution occurs across ranks 0 to 3.
Thus, users of ``/home/backups`` may be affected by noisy neighbors unnecessarily.
By distributing the two directories to MDS rank 0,1 group and MDS rank 2,3 group
respectively through the ``ceph.dir.bal.mask``, metadata service can be provided
without affecting the performance of other concurrent workloads.

This option can be set via:

::

    setfattr -n ceph.dir.bal.mask -v 1,2 /home/images
    setfattr -n ceph.dir.bal.mask -v 3,4 /home/backups

``/homes/images`` and ``/home/backups`` are distributed within mds rank 1~2 and 3~4, respectively.

Similar to ``ceph.dir.pin``, the ``ceph.dir.bal.mask`` is also inherited from its closest parent.
This manner involves configuring the ``ceph.dir.bal.mask`` for a directory, which consequently

> impacts all its subordinate items. However, the parent ``ceph.dir.bal.mask`` can be overriden

Author: Fixed.

by setting the child directory’s value. For example:

::

    mkdir -p /home/images/jpg
    setfattr -n ceph.dir.bal.mask -v 1,2 /home/images

>     # /home/images and /home/images/jpg are now freely distributed within mds rank 1 and 2

Author: Fixed.

    setfattr -n ceph.dir.bal.mask -v 3,4 /home/images/jpg

>     # /home/images/jpg are distributed within mds rand 3 and 4

Author: Fixed.

The option can be unset via:

::

>     setfattr -n ceph.dir.bal.mask -v -1 /home/images/jpg

Author: Fixed.

The value of the ``ceph.dir.bal.mask`` for  ``/home/images/jpg`` is unset and replaced with their inherited value from
its nearest parent. If the ``ceph.dir.bal.mask`` for /home/images is configured with
a valid value, ``/home/images/jpg`` will be distributed according to its parent’s settings.

This can override the mdsmap’s ``bal_rank_mask``. For example:

::

    cephfs fs set cephfs bal_rank_mask 0xf

Initially, the balancer dynamically partitions the file system within MDS rank
0 to 3 because the ``bal_rank_mask`` is set to ``0xf``.

If the ``ceph.dir.bal.mask`` for the root ``/`` directory is set to ``0,1,2,3,4,5``,
the ``bal_rank_mask`` will be overridden. In this way, the balancer dynamically

> distributes the unpinned subtrees of the root ``/`` directory from ``0`` to ``5`` mds ranks.

Author: ```bal_rank_mask``` does not need to be deprecated. If detailed tuning is required according to the user's preference, ```ceph.dir.bal.mask``` xattr can be used. Otherwise, ```bal_rank_mask``` can be used for simplicity.

> distributes the unpinned subtrees of the root ``/`` directory from ``0`` to ``5`` mds ranks.

Reviewer: Perhaps add a note about that, literally:

```
.. note:: Users may prefer to use the file system setting `bal_rank_mask` ...
```

> distributes the unpinned subtrees of the root ``/`` directory from ``0`` to ``5`` mds ranks.

Author: Added. https://github.com/ceph/ceph/compare/9b04cd83fb615464888de665b9933d37724d4a28..ac1ab14f09a789361f9a5c72e448ae85f5b55fa8#diff-cc9cf0d49df0c5644ecb9c98d1f7ac1d307b05837bfdabc6761c65691c6f8dc1R407

::

    setfattr -n ceph.dir.bal.mask -v 0,1,2,3,4,5 /

``ceph.dir.bal.mask`` overrides parent ``ceph.dir.pin`` and vice versa. For example:

::

    mkdir -p /home/images/jpg
    setfattr -n ceph.dir.pin -v 1 /home/images
    setfattr -n ceph.dir.bal.mask -v 2,3 /home/images/jpg

The ``/home/images`` directory will be pinned to rank 1, while the ``/home/images/jpg``

> directory will be dynamically split into rank 2 and 3.

Author: Fixed.

> **Restrictions**: Since the inode of the root directory is anchored in MDS rank0,

Author: Fixed.

> it's reuiqred to include MDS rank 0 when adjusting the ceph.dir.rank.mask for the ``/`` directory.

Author: Fixed.

.. prompt:: bash #

    ceph fs set <fs_name> bal_rank_mask 0x0

Dynamically partitionning directory trees with vxattr on a specific ranks
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The ``bal_rank_mask`` file system setting (see also :ref:`intro-bal-rank-mask`)
can be overridden by configuring the ``ceph.dir.bal.mask`` vxattr.
Notably, the ``ceph.dir.bal.mask`` offers the capability to isolate particular subdirectories
into dedicated groups of MDS ranks, whereas bal_rank_mask only allows isolation
at the file system level, such as the root directory ``/``. It serves as a valuable
tool for finely tuning and enhancing MDS performance in various scenarios.

One scenario is when dealing with a large directory that proves challenging
for a single MDS rank to handle efficiently. For example, the ``/usr`` directory
contains many directories for multiple users. In most cases, static pinning is
simply used to distribute subdirectories of ``/usr`` across multiple ranks. However,
the ``/usr/share`` directory presents a unique situation with many subdirectories,

> each containing large number of files, often exceeding a million files. With
> this example with ``/usr/share/images``, processing them in one MDS is inappropriate
> due to insufficient MDS resources like MDS cache memory. Therefore,
> the ceph.dir.bal.mask can dynamically balance the workload of this large
> directory within a specific subset of ranks.

Reviewer: ```suggestion
each containing large number of files, often exceeding a million files. For
example, consider``/usr/share/images`` with several million images. Processing this directory on one MDS may be inappropriate
due to insufficient MDS resources like MDS cache memory. Therefore,
the ceph.dir.bal.mask could dynamically balance the workload of this large
directory within a specific subset of ranks.
```

and ``/usr/share/backups`` within the file system.

.. note:: The example here uses the /usr/share directory.
        Depending on the user's preference, it can be applied
        to various directories such as /usr/share and /mnt/cephfs.

While the file system setting ``bal_rank_mask`` isolates the entire ``/`` directory to
specific ranks, it can affect performance due to each other's migration overhead.
For example, if the file system setting``bal_rank_mask`` is set to ``0xf`` and large directories
like ``/usr/share/images`` and ``/usr/share/backups`` exist, the load on ``/usr/share/images``
instantaneously increases and metadata distribution occurs across ranks 0 to 3.
Thus, users of ``/usr/share/backups`` may be affected by noisy neighbors unnecessarily.
By distributing the two directories to MDS rank 0,1 group and MDS rank 2,3 group
respectively through the ``ceph.dir.bal.mask``, metadata service can be provided
without affecting the performance of other concurrent workloads.

This option can be set via:

::

    setfattr -n ceph.dir.bal.mask -v 1,2 /usr/share/images
    setfattr -n ceph.dir.bal.mask -v 3,4 /usr/share/backups

> .. note:: The balance_automate setting should be `true` for this option to be turned on.

Reviewer: ```suggestion
.. note:: The balance_automate setting must be `true` for `ceph.dir.bal.mask` to have any effect.
```

0 to 3 because the ``bal_rank_mask`` is set to ``0xf``.

If the ``ceph.dir.bal.mask`` for the root ``/`` directory is set to ``0,1,2,3,4,5``,
the ``bal_rank_mask`` will be overridden. In this way, the balancer dynamically
distributes the unpinned subtrees of the root ``/`` directory from ``0`` to ``5`` mds ranks.

.. note:: Users may prefer to use the file system setting `bal_rank_mask` for simplicity.

::

    setfattr -n ceph.dir.bal.mask -v 0,1,2,3,4,5 /

``ceph.dir.bal.mask`` overrides parent ``ceph.dir.pin`` and vice versa. For example:

::

    mkdir -p /usr/share/images/jpg
    setfattr -n ceph.dir.pin -v 1 /usr/share/images
    setfattr -n ceph.dir.bal.mask -v 2,3 /usr/share/images/jpg

The ``/usr/share/images`` directory will be pinned to rank 1, while the ``/usr/share/images/jpg``
directory may be dynamically split into rank 2 and 3.

While `ceph.dir.bal.mask` functions similarly to `ceph.dir.pin` when it has only one mds rank,
its distinct advantage lies in its flexibility.
Unlike `ceph.dir.pin`, `ceph.dir.bal.mask` can be expanded to multiple mds ranks or reduced to fewer mds ranks.
Therefore, in scenarios requiring expansion to multiple mds ranks,

> it is recommended to employ `ceph.dir.bal.mask`.

Reviewer: On the other hand, if the admin wants the automated balancer to be off, then `ceph.dir.pin` or ephemeral pinning is the way to distribute load.

::

    # Same way as setfattr -n ceph.dir.pin -v 1 /usr/share/images
    setfattr -n ceph.dir.bal.mask -v 1 /usr/share/images

    # If expansion is necessary, multiple mds ranks can be manipulated.
    setfattr -n ceph.dir.bal.mask -v 1,2,3 /usr/share/images

If the max_mds number shrinks, the subtree should move to mds rank 0.
Therefore, max_mds should be adjusted carefully because MDS performance may decrease.

::
    # The file system is operating with mds ranks 0 to 3.
    ceph fs set cephfs max_mds 4

    # The subtree is moved among mds ranks 2 and 3.
    setfattr -n ceph.dir.bal.mask -v 2,3 /usr/share/images

    # The file system is operating with mds ranks 0 and 1.
    ceph fs set cephfs max_mds 2

    # The subtree is now moved to mds rank 0.

> **Restrictions**: Since the inode of the root directory is anchored to MDS rank 0,

Reviewer: ```suggestion
**Restrictions**: Since the inode of the root directory is always implicitly pinned to MDS rank 0,
```

.. prompt:: bash #

    ceph fs set <fs_name> bal_rank_mask 0x0

Dynamically partitioning directory trees onto specific ranks
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The ``bal_rank_mask`` file system setting (see also :ref:`intro-bal-rank-mask`)
can be overridden by configuring the ``ceph.dir.bal.mask`` vxattr.
Notably, the ``ceph.dir.bal.mask`` offers the capability to isolate particular subdirectories
into dedicated groups of MDS ranks, whereas bal_rank_mask only allows isolation

> at the file system level, such as the root directory ``/``. It serves as a valuable

Author: Updated as suggested

One scenario is when dealing with a large directory that proves challenging
for a single MDS rank to handle efficiently. For example, the ``/usr`` directory
contains many directories for multiple users. In most cases, static pinning is
simply used to distribute subdirectories of ``/usr`` across multiple ranks. However,
the ``/usr/share`` directory presents a unique situation with many subdirectories,
each containing large number of files, often exceeding a million files. For
example, consider ``/usr/share/images`` with several million images. Processing this directory on one MDS may be inappropriate
due to insufficient MDS resources like MDS cache memory. Therefore,
the ceph.dir.bal.mask could dynamically balance the workload of this large
directory within a specific subset of ranks.

The ``ceph.dir.bal.mask`` can also be a useful option for fine-tuning performance,
particularly in scenarios involving large directories like ``/usr/share/images``
and ``/usr/share/backups`` within the file system.

.. note:: The example here uses the /usr/share directory.
        Depending on the user's preference, it can be applied

>         to various directories such as /usr/share and /mnt/cephfs.

Reviewer 2: double backticks?

While the file system setting ``bal_rank_mask`` isolates the entire ``/`` directory to

> specific ranks, it can affect performance due to each other's migration overhead.

Author: Updated as suggested

For example, if the file system setting``bal_rank_mask`` is set to ``0xf`` and large directories
like ``/usr/share/images`` and ``/usr/share/backups`` exist, the load on ``/usr/share/images``
instantaneously increases and metadata distribution occurs across ranks 0 to 3.
Thus, users of ``/usr/share/backups`` may be affected by noisy neighbors unnecessarily.
By distributing the two directories to MDS rank 0,1 group and MDS rank 2,3 group
respectively through the ``ceph.dir.bal.mask``, metadata service can be provided
without affecting the performance of other concurrent workloads.

This option can be set via:

::

    setfattr -n ceph.dir.bal.mask -v 1,2 /usr/share/images
    setfattr -n ceph.dir.bal.mask -v 3,4 /usr/share/backups

> .. note:: The balance_automate setting must be `true` for `ceph.dir.bal.mask` to have any effect.

Author: Updated as suggested

> ``/usr/share/images`` and ``/usr/share/backups`` are distributed within mds rank 1~2 and 3~4, respectively.

Author: Updated as suggested

Similar to ``ceph.dir.pin``, the ``ceph.dir.bal.mask`` is also inherited from its closest parent.

> This manner involves configuring the ``ceph.dir.bal.mask`` for a directory, which consequently

Author: Updated as suggested

impacts all its descendants. However, the parent ``ceph.dir.bal.mask`` can be overridden
by setting the child directory’s value. For example:

> ::

Reviewer 2: @bluikko Ditto wrt this block and those below

0 to 3 because the ``bal_rank_mask`` is set to ``0xf``.

If the ``ceph.dir.bal.mask`` for the root ``/`` directory is set to ``0,1,2,3,4,5``,
the ``bal_rank_mask`` will be overridden. In this way, the balancer dynamically
distributes the unpinned subtrees of the root ``/`` directory from ``0`` to ``5`` mds ranks.

.. note:: Users may prefer to use the file system setting `bal_rank_mask` for simplicity.

::

    setfattr -n ceph.dir.bal.mask -v 0,1,2,3,4,5 /

``ceph.dir.bal.mask`` overrides parent ``ceph.dir.pin`` and vice versa. For example:

::

    mkdir -p /usr/share/images/jpg
    setfattr -n ceph.dir.pin -v 1 /usr/share/images
    setfattr -n ceph.dir.bal.mask -v 2,3 /usr/share/images/jpg

The ``/usr/share/images`` directory will be pinned to rank 1, while the ``/usr/share/images/jpg``
directory may be dynamically split into rank 2 and 3.

While `ceph.dir.bal.mask` functions similarly to `ceph.dir.pin` when it has only one mds rank,
its distinct advantage lies in its flexibility.
Unlike `ceph.dir.pin`, `ceph.dir.bal.mask` can be expanded to multiple mds ranks or reduced to fewer mds ranks.
Therefore, in scenarios requiring expansion to multiple mds ranks,

> it is recommended to employ `ceph.dir.bal.mask`.

Reviewer 2: MDS uppercase
Double backticks around 5 literals in the above

::

    # Same way as setfattr -n ceph.dir.pin -v 1 /usr/share/images
    setfattr -n ceph.dir.bal.mask -v 1 /usr/share/images

    # If expansion is necessary, multiple mds ranks can be manipulated.
    setfattr -n ceph.dir.bal.mask -v 1,2,3 /usr/share/images

If the max_mds number shrinks, the subtree should move to mds rank 0.
Therefore, max_mds should be adjusted carefully because MDS performance may decrease.

::
    # The file system is operating with mds ranks 0 to 3.
    ceph fs set cephfs max_mds 4

    # The subtree is moved among mds ranks 2 and 3.
    setfattr -n ceph.dir.bal.mask -v 2,3 /usr/share/images

    # The file system is operating with mds ranks 0 and 1.
    ceph fs set cephfs max_mds 2

    # The subtree is now moved to mds rank 0.

**Restrictions**: Since the inode of the root directory is always implicitly pinned to MDS rank 0,

> it's required to include MDS rank 0 when adjusting the ceph.dir.rank.mask for the ``/`` directory.

Author: Updated as suggested
